aiomisc/circuit_breaker.py

```diff
@@ -0,0 +1,326 @@
+import asyncio
+import threading
+import typing
+from collections import Counter, deque
+from contextlib import contextmanager
+from enum import IntEnum, unique
+from functools import wraps
+from random import random
+
+from aiomisc.utils import awaitable
+
+
+TimeType = typing.Union[int, float]
+StatisticType = typing.Deque[typing.Tuple[int, Counter]]
+
+
+@unique
+class CounterKey(IntEnum):
+    FAIL = 0
+    OK = 1
+    TOTAL = 2
+
+
+@unique
+class CircuitBreakerStates(IntEnum):
+    PASSING = 0
+    BROKEN = 1
+    RECOVERING = 2
+
+
+class CircuitBroken(Exception):
+    pass
+
+
+class CircuitBreaker:
+    __slots__ = (
+        "_broken_time",
+        "_error_ratio",
+        "_exceptions",
+        "_lock",
+        "_loop",
+        "_passing_time",
+        "_recovery_at",
+        "_recovery_ratio",
+        "_recovery_time",
+        "_response_time",
+        "_state",
+        "_statistic",
+        "_stuck_until",
+    )
+
+    BUCKET_COUNT = 10
+    # Thresholds when state will be changed
+
+    # * RECOVER state will be changed to BROKEN
+    #   when error ratio will be greater or equal then
+    #   RECOVER_BROKEN_THRESHOLD
+    RECOVER_BROKEN_THRESHOLD = 0.5
```

Comment:
Maybe that should be lower in some cases. At least it will be good to allow customizing it.

Reply:
May be, but on practice it means how fast return state to broken when still failed. I guess this value might be changed extremely not often